app/controllers/cases_controller.rb

```diff
@@ -278,37 +285,173 @@ def destroy
     end
   end
 
-  def new_response_upload
-    flash[:action_params] = request.query_parameters['mode']
-    authorize_upload_response_for_action @case, flash[:action_params]
-    @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
+  def approve
+    authorize @case
+
+    @case = @case.decorate
+  end
+
+  def execute_approve
+    authorize @case, :approve?
+
+    case_approval_service = CaseApprovalService.new(
+      user: current_user,
+      kase: @case,
+      bypass_params: BypassParamsManager.new(params)
+    )
+    case_approval_service.call
+
+    if case_approval_service.result == :ok
+      current_team = CurrentTeamAndUserService.new(@case).team
+      if @case.ico?
+        flash[:notice] = t('notices.case/ico.case_cleared')
+      else
+        flash[:notice] = t('notices.case_cleared', team: current_team.name,
+                                                   status: I18n.t("state.#{@case.current_state}").downcase)
+      end
+      redirect_to case_path(@case)
+    else
+      flash[:alert] = case_approval_service.error_message
+      @case = @case.decorate
+      render :approve
+    end
   end
 
   def upload_responses
-    authorize_upload_response_for_action @case, flash[:action_params]
+    authorize @case
+
+    @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
+    @case = @case.decorate
+  end
+
+  def upload_responses_action
+    authorize @case, :upload_responses?
 
-    bypass_params_manager = BypassParamsManager.new(params)
     rus = ResponseUploaderService.new(
-      @case, current_user, bypass_params_manager, flash[:action_params]
+      kase: @case,
+      current_user: current_user,
+      action: 'upload',
+      uploaded_files: params[:uploaded_files],
+      upload_comment: params[:upload_comment],
+      is_compliant: false,
+      bypass_further_approval: false,
+      bypass_message: nil
     )
     rus.upload!
+
+    @s3_direct_post = S3Uploader.s3_direct_post_for_case(@case, 'responses')
+
+    case rus.result
+    when :blank
+      flash.now[:alert] = t('alerts.response_upload_blank?')
+      render :upload_responses
+    when :error
```

Comment:
As above comment, possible to use shorthand i.e. `render :upload_responses, alert: ....` - also is the `.now` needed? if so a comment would help understanding.

Reply:
After discussion with @starswan the `.now` is needed to ensure the flash message is displayed in the current `render` as per documentation https://guides.rubyonrails.org/action_controller_overview.html#the-flash